reactor-extra/src/main/java/reactor/retry/RetryBuilder.java

```diff
@@ -0,0 +1,172 @@
+import java.util.function.Predicate;
+
+import org.reactivestreams.Publisher;
+
+import reactor.core.publisher.EmitterProcessor;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * Retry utility to control retry or repeat of {@link Flux} or @link {@link Mono} using
+ * <ul>
+ *   <li>{@link Flux#retryWhen(Function)} </li>
+ *   <li>{@link Flux#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#retryWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhen(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(Function)} </li>
+ *   <li>{@link Mono#repeatWhenEmpty(int, Function)} </li>
+ * </ul>
+ *
+ * This class enables commonly used retry/repeat patterns:
+ * <ul>
+ *   <li>Configured number of retries without backoff</li>
+ *   <li>Configured number of retries with fixed backoff interval</li>
+ *   <li>Exponential backoff with timeout</li>
+ *   <li>Retriable exception configuration that can be used with/without backoff</li>
+ *   <li>Non-retriable exception configuration that can be used with/without backoff</li>
+ *   <li>doOnRetry to perform any rollback operation before performing retry</li>
+ * </ul>
+ *
+ * This class does not support fully flexible functions which can be done just as easily on {@link Flux#retryWhen(Function)} etc.
+ * <ul>
+ *   <li>Generic predicate for retry decision</li>
+ *   <li>Flexible delay based on context</li>
+ * </ul>
+ *
+ */
+public class RetryBuilder {
+
+	static final Duration NO_RETRY = Duration.ofSeconds(-1);
+
+	final Predicate<Throwable> retryPredicate;
+	Function<Long, Duration> delayCalculator;
+	Consumer<RetryContext> onRetry;
+
+	RetryBuilder(Predicate<Throwable> retryPredicate) {
+		this.retryPredicate = retryPredicate;
+		this.delayCalculator = iteration -> Duration.ZERO; // No backoff by default
+		this.onRetry = r -> {};
+	}
+
+	public static RetryBuilder any() {
+		return new RetryBuilder(e -> true);
+	}
+
+	@SafeVarargs
+	public static RetryBuilder anyOf(Class<? extends Throwable>... retriableExceptions) {
+		return new RetryBuilder(e -> {
+			for (Class<? extends Throwable> clazz : retriableExceptions) {
+				if (clazz.isInstance(e))
+					return true;
+			}
+			return false;
+		});
+	}
+
+	@SafeVarargs
+	public static RetryBuilder allBut(final Class<? extends Throwable>... nonRetriableExceptions) {
+		return new RetryBuilder(e -> {
+			for (Class<? extends Throwable> clazz : nonRetriableExceptions) {
+				if (clazz.isInstance(e))
+					return false;
+			}
+			return true;
+		});
+	}
+
+	public RetryBuilder noBackoff(int maxAttempts) {
+		this.delayCalculator = iteration -> iteration < maxAttempts ? Duration.ZERO : NO_RETRY;
+		return this;
+	}
+
+	public RetryBuilder fixedBackoff(int maxAttempts, Duration backoffInteval) {
+		this.delayCalculator = iteration -> iteration < maxAttempts ? backoffInteval : NO_RETRY;
+		return this;
+	}
+
+	public RetryBuilder exponentialBackoff(Duration firstBackoff, Duration maxBackoff, Duration timeout) {
```

Comment:
looks like exponential backoff cannot be limited to a maximum number of attempts?